taskcluster/app_services_taskgraph/transforms/beetmover.py

```diff
@@ -2,74 +2,183 @@
-def build_upstream_artifacts(config, tasks):
+def resolve_keys(config, tasks):
     for task in tasks:
-        module_info = task["attributes"]["buildconfig"]
-        version = get_version(config.params)
-        publications = module_info["publications"]
+        for key in ("worker.action",):
+            resolve_keyed_by(
+                task, key, item_name=task["name"], level=config.params["level"]
+            )
+        yield task
+
+
+def _build_upstream_artifacts(config, task):
+    upstream_artifacts = []
+
+    for label, dep in config.kind_dependencies_tasks.items():
+        if label not in task["dependencies"].values():
+            continue
+
+        paths = sorted(
+            artifact["name"] for artifact in dep.attributes.get("release-artifacts", [])
+        )
+
+        if paths:
+            upstream_artifacts.extend(
+                [
+                    {
+                        "taskId": {"task-reference": f"<{dep.kind}>"},
+                        "taskType": "repackage",
+                        "paths": paths,
+                    }
+                ]
+            )
+
+    return upstream_artifacts
+
 
-        worker_definition = {"upstream-artifacts": [{
+def _build_maven_upstream_artifacts(config, task):
+    module_info = task["attributes"]["buildconfig"]
+    version = get_version(config.params)
+    publications = module_info["publications"]
+
+    return [
+        {
             "taskId": {"task-reference": "<module-build>"},
             "taskType": "build",
-            "paths": (publications_to_artifact_paths(version, publications,
-                                                     ("", ".sha1", ".md5"))),
-        }, {
+            "paths": (
+                publications_to_artifact_paths(
+                    version, publications, ("", ".sha1", ".md5")
+                )
+            ),
+        },
+        {
             "taskId": {"task-reference": "<signing>"},
             "taskType": "signing",
             "paths": (publications_to_artifact_paths(version, publications, (".asc",))),
-        }]}
+        },
+    ]
+
 
+@transforms.add
+def build_upstream_artifacts(config, tasks):
+    for task in tasks:
         task.setdefault("worker", {})
-        task["worker"].update(worker_definition)
+        if task["worker"]["action"] == "push-to-maven":
+            task["worker"]["upstream-artifacts"] = _build_maven_upstream_artifacts(
+                config, task
+            )
+
+        else:
+            task["worker"]["upstream-artifacts"] = _build_upstream_artifacts(
+                config, task
+            )
+
         yield task
 
 
-@transforms.add
-def build_artifact_map(config, tasks):
-    for task in tasks:
-        module_info = task["attributes"]["buildconfig"]
-        version = get_version(config.params)
+def _build_artifact_map(config, task):
+    version = get_version(config.params)
+
+    destination_paths = []
+    if config.params.get("shipping_phase") == "promote":
```

Comment:
I don't think it can sensibly be a single task.  Per the snowman model we want the ship graph to be a superset of the promote graph, so can't have the "same" tasks (same label etc) get tweaked depending on the phase.

Reply:
Ah ok, that makes sense. I think it still works technically (at least it seems to with VPN), but I can see why it might cause confusion to an observer.

Reply:
Ah right, that's what would happen.. Though arguably in the case of beetmover that would kind of reflect reality better as that release never had an associated candidate :p

But yeah, I agree with you.. Updated patch incoming shortly.